### PR TITLE
New option to add submissions from other nodes

### PR DESCRIPTION
**Pull request by katzmo**

### CODE REVIEW

File: src/Source/WebformComponentSum.php [hunk 1]
```diff
@@ -38,28 +39,35 @@ public function __construct($entity, $instance) {
...
-WHERE
-  (n.nid=:nid OR ((n.nid=:tnid OR n.tnid=:tnid) AND :tnid>0)) AND wc.form_key=:fkey
-EOSQL;
-    $args = array(
-      ':nid' => $entity->nid,
-      ':tnid' => $entity->tnid,
-      ':fkey' => $item['options']['source']['form_key'],
-    );
-    return db_query($sql, $args)->fetchField();
+    $nids = array($entity->nid);
```
Comment (torotil) [inline src/Source/WebformComponentSum.php]: Please use the [] array syntax for new code.

File: src/Source/WebformComponentSum.php [hunk 2]
```diff
@@ -38,28 +39,35 @@ public function __construct($entity, $instance) {
...
+    if (isset($item['options']['source']['add_nids'])
+        && $add_nids = $item['options']['source']['add_nids']) {
+      array_push($nids, ...$add_nids);
+    }
+    $form_key = $item['options']['source']['form_key'];
+
+    $q_nids = db_select('node', 'n');
+    $q_nids->leftJoin('node', 'nt', 'n.tnid>0 AND nt.tnid=n.tnid');
+    $q_nids->addExpression('COALESCE(nt.nid, n.nid)', 'nid');
+    $q_nids->condition('n.nid', $nids);
```
Comment (torotil) [inline src/Source/WebformComponentSum.php]: Nothing that needs changing, but still worth mentioning. All query methods return the query object so they are chainable — except for the joins. So you could write this as:

```php
$q_nids = db_select('node', 'n');
$nt = $q_nids->leftJoin('node', 'nt', 'n.tnid>0 AND nt.tnid=n.tnid');
$q_nids->addExpression("COALESCE({$nt}.nid, n.nid)", 'nid')
  ->condition('n.nid', $nids);
```

File: src/Source/WebformComponentSum.php [hunk 3]
```diff
@@ -68,6 +76,41 @@ public function widgetForm($item) {
       '#desription' => t('All values with this form key are summed up to get the current value for the progress bar.'),
       '#default_value' => $source_options['form_key'],
     );
+    // add_nids
+    $source_options = isset($item['options']['source']) ? $item['options']['source'] : array();
+    $source_options += array('add_nids' => '');
```
Comment (torotil) [inline src/Source/WebformComponentSum.php]: Please add all the default config in one place at the beginning of the function. You might want to use `\Drupal\little_helpers\ArrayConfig::mergeDefaults()` to pack this into a single statement.

File: src/Source/WebformComponentSum.php [hunk 4]
```diff
@@ -68,6 +76,41 @@ public function widgetForm($item) {
...
+  function widgetValidate($element, &$form_state, $form) {
+    $item = &$form_state['values'];
+    foreach ($element['#parents'] as $key) {
+      $item = &$item[$key];
+    }
+    $nids = &$item['add_nids'];
+    $parts = array_filter(explode(',', $nids));
+    $nids = array();
+    foreach ($parts as $p) {
+      dpm($p, 'loop');
```
Comment (torotil) [inline src/Source/WebformComponentSum.php]: No `dpm()`s please ;)

File: src/Source/WebformSubmissionCount.php [hunk 5]
```diff
@@ -24,28 +24,72 @@ public static function forField($entity, $field, $instance) {
...
+      '#default_value' => $source_options['add_nids'],
+    );
+    $form['#element_validate'][] = array($this, 'widgetValidate');
+    return $form;
+  }
+
+  /**
+   * Element validate callback for @see widgetForm().
+   */
+  function widgetValidate($element, &$form_state, $form) {
```
Comment (torotil) [inline src/Source/WebformSubmissionCount.php]: A lot of this functionality seems to be duplicated in both plugins. We might want to move the functionality to it’s own class and use that in both of the plugins (composition or inheritance).

File: src/Source/WebformSubmissionCount.php [hunk 6]
```diff
@@ -24,28 +24,72 @@ public static function forField($entity, $field, $instance) {
...
+   *   and on additional nodes provided via the field widget,
+   *   and all their translations.
    */
   public function getValue($item) {
     $entity = $this->entity;
-    $q = db_select('node', 'n');
+    $nids = array($entity->nid);
+    if (isset($item['options']['source']['add_nids'])
+        && $add_nids = $item['options']['source']['add_nids']) {
+      array_push($nids, ...$add_nids);
```
Comment (torotil) [inline src/Source/WebformSubmissionCount.php]: [The ... syntax](http://php.net/manual/en/functions.arguments.php#functions.variable-arg-list) was introduced in PHP 5.6. Currently we only require PHP 5.4 (new-style array syntax). We could bump the requirement to PHP 5.6 but should do so explicitly.

**Review by torotil — DISMISSED** [review]
Thanks. Still a it of refactoring and polishing needed.

**Review by alex-mo — APPROVED** [review]

**Pull request merged**